api/envoy/config/bootstrap/v3/bootstrap.proto

```diff
@@ -186,6 +187,9 @@ message Bootstrap {
   // Specifies optional bootstrap extensions to be instantiated at startup time.
   // Each item contains extension specific configuration.
   repeated core.v3.TypedExtensionConfig bootstrap_extensions = 21;
+
+  // Optional loading of custom socket interfaces and configuration of the default one
+  core.v3.SocketInterfacesConfig socket_interfaces = 22;
```

Comment:
I don't think you really need the new field, you should be able to roll this PR entirely into a bootstrap_extensions above? i.e. you should be able to register `SocketInterfaceFactory` as `BootstrapExtensionFactory` and load socket interface there.

Reply:
That was the first version of the PR but unfortunately the bootstrap extensions are loaded after the `ListenerManagerImpl` is created and initialized, so after the listen sockets are allocated. If you think it's fine to move the bootstrap extensions [init code](https://github.com/envoyproxy/envoy/blob/master/source/server/server.cc#L495) to [here](https://github.com/envoyproxy/envoy/blob/master/source/server/server.cc#L413) we should be able to use them.

Reply:
Also, @lizan if you have the time, do let us know what you think of [this]!(https://docs.google.com/document/d/1y_UOJjoWw70Wxb9TQ8Il7KdudjT7EonoSSY8W9E6fGo/edit?usp=sharing)

Reply:
> That was the first version of the PR but unfortunately the bootstrap extensions are loaded after the `ListenerManagerImpl` is created and initialized, so after the listen sockets are allocated. If you think it's fine to move the bootstrap extensions [init code](https://github.com/envoyproxy/envoy/blob/master/source/server/server.cc#L495) to [here](https://github.com/envoyproxy/envoy/blob/master/source/server/server.cc#L413) we should be able to use them.

Yeah that would be fine. Since there is nothing else using bootstrap extension (there will be in wasm but not upstreamed here) now, just go ahead and move it ahead.

When I added bootstrap extensions I realized there will be needs for different initialization phase, so we might want add that option later.

Reply:
Done! If you can, do check if I'm using bootstrap extensions the way you expect them to be used. Also, as per @mattklein123's comment, if you know of any way that we could simplify/avoid the building of a hash map in the SocketInterfacesSingleton, we could use it to simplify the current logic. That is, config handed to factory registered with `REGISTER_FACTORY` which creates and registers `SocketInterface` instances in the `SocketInterfacesSingleton`.

Reply:
The registry is itself a static singleton. Can't we just use it directly to get a factory and then create a socket?

Reply:
As far as I can tell, that would be equivalent to using the bootstrap extension config to configure the `SocketInterface` factories, as opposed to using it to generate "customized" `SocketInterface` instances. For scenarios when no of little customization is needed, this should work without issue. In case of VCL, I think this would work, but that's probably not the most generic configuration use case.  

As a side note, the objects to be generated by the factories, since all classes that derive `BootstrapExtensionFactory` must override `createBootstrapExtension`, will most probably be just dummies (and will end up stored in `bootstrap_extensions_` vector). But, I guess, that shouldn't be a problem.

Reply:
Yeah I guess if we are going to use bootstrap extension for this we will need a separate registration system to store the factories. @lizan ?

Reply:
Nested factories were always doable, (e.g. http filters facotry registered by HCM, itself registered as network filter factory.) I don't think you need a hash map for this purpose.

> As a side note, the objects to be generated by the factories, since all classes that derive `BootstrapExtensionFactory` must override `createBootstrapExtension`, will most probably be just dummies (and will end up stored in `bootstrap_extensions_` vector). But, I guess, that shouldn't be a problem.

The `BootstrapExtension` is an opaque object, which should be the singleton if you expect the lifecycle of the singleton lives with the server instance.

Reply:
@lizan I think the nested HCM case is a bit different. Here we need to basically register a socket factory, and I think what is being proposed is doing that within a bootstrap extenstion?

Reply:
@mattklein123 is the latest update closer to what you are thinking?

Reply:
@florincoras IIUC, what you want to:

1. register multiple SocketInterfaceFactories with config (via bootstrap extensions)
2. specify a default socket interface in bootstrap.proto
3. override socket interface to use in addresses.

Is that right? If so you should be able to use the bootstrap extension name to lookup the factory instead of using a flat hash map?

Reply:
@lizan In this PR, only the first 2, the 3rd will involve changes to addresses, so it'll be a separate PR. 

With the latest change, the `SocketInterface` implementations also derive `BootstrapExtensionFactory` so the lookup is done like [this](https://github.com/envoyproxy/envoy/pull/11630/files#diff-d28ab6f8cef00963ff99f85579f7308dR46). And [here](https://github.com/envoyproxy/envoy/pull/11630/files#diff-5c03976b094abde0bb70c44295d1cd4bR11) is an example of the changes needed for a `SocketInterface` implementation.

Reply:
@lizan, if you get a chance to take a look at the latest commit, do let me know if the approach seems okay to you.

Reply:
Looks it is in right direction. Yeah I agree 3 shouldn't be in this PR, but just to understand the context.